src/topology/algebra/weak_dual_topology.lean

```diff
@@ -0,0 +1,87 @@
+/-
+Copyright (c) 2021 Kalle Kytölä. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kalle Kytölä and Heather Macbeth
+-/
+import tactic
+import topology.algebra.module
+
+noncomputable theory
+open filter
+open_locale topological_space
+
+section weak_star_topology
+
+/-!
+### Weak star topology on duals of topological modules
+In this section, we define the weak-* topology on duals of suitable topological modules `E` over
+suitable topological semirings `𝕜`. The (weak) dual consists of continuous linear functionals
+`E →L[𝕜] 𝕜` from `E` to scalars `𝕜`. The weak-* topology is the coarsest topology on this dual
+`weak_dual 𝕜 E := (E →L[𝕜] 𝕜)` w.r.t. which the evaluation maps at all `z : E` are continuous.
+
+The weak dual is a module over `𝕜` if the semiring `𝕜` is commutative.
+-/
+
+variables (𝕜 : Type*) [topological_space 𝕜] [semiring 𝕜]
+variables [has_continuous_add 𝕜] [has_continuous_mul 𝕜]
+variables (E : Type*) [topological_space E] [add_comm_monoid E] [has_continuous_add E]
+variables [module 𝕜 E] [has_continuous_smul 𝕜 E]
+
+/-- The (weak) dual of a topological module `E` over a topological semiring `𝕜` consists of
+continuous linear functionals from `E` to scalars `𝕜`. It is a type synonym with the original
+dual, but will be equipped with a different topology. -/
+def weak_dual : Type* := (E →L[𝕜] 𝕜)
+
+instance [inhabited 𝕜] : inhabited (weak_dual 𝕜 E) :=
+{ default := { to_fun := (λ (x : E) , 0),
+               map_add' := by simp only [add_zero, forall_const],
+               map_smul' := by simp only [forall_const, smul_zero],
+               cont := continuous_const, }, }
+
+instance weak_dual.has_coe_to_fun :
+  has_coe_to_fun (weak_dual 𝕜 E) := ⟨_, (λ (x' : weak_dual 𝕜 E), (λ (x : E), x'.to_fun x)) ⟩
+
+instance weak_dual.topology :
+  topological_space (weak_dual 𝕜 E) :=
+topological_space.induced (λ (x' : (weak_dual 𝕜 E)), (λ (z : E), x' z)) Pi.topological_space
```

Comment:
In many cases you can get away with fewer parentheses than you've used -- for example, here
```suggestion
topological_space.induced (λ x' : weak_dual 𝕜 E, λ z : E, x' z) Pi.topological_space
```
works.  Feel free either to aggressively strip away parentheses throughout the file, or to keep those which you feel aid readability :)